Exception based scheduling

br648 opened the pull request:
### Checklist

- [ ] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [ ] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [ ] The description lists all applicable issues this PR seeks to resolve
- [ ] The description lists any configuration setting(s) that differ from the default settings
- [ ] All tests and CI builds passing

### Description

This PR includes the exception based scheduling updates made to GTFS-lib (https://github.com/conveyal/gtfs-lib/pull/385) and a rector of feed merge to accommodate multiple table references.

Commit:
feat(Bumped GTFS lib version and put a temp fix in place for multi table refs):

Commit:
refactor(MergeLineContext): Updated to use multi table refs

br648 commented:
@binh-dam-ibigroup before i dive into a fix (to replace the temp fix that doesn't work) do you have any thoughts on how to approach this which might save me a bit of time?!

Commit:
refactor(Merge line context update): Refactored to work with multiple table references

Commit:
refactor(Addressed line spacing and updated comments):

binh-dam-ibigroup commented on src/main/java/com/conveyal/datatools/manager/jobs/feedmerge/MergeLineContext.java:
```diff
@@ -237,42 +237,104 @@ public void startNewRow() throws IOException {
+            case TRIP_SERVICE_ID_KEY:
+                boolean isCalendarServiceId = mergeFeedsResult.calendarServiceIds.contains(fieldContext.getValueToWrite());
+                boolean isCalendarDatesServiceId = mergeFeedsResult.calendarDatesServiceIds.contains(fieldContext.getValueToWrite());
+                if (isCalendarServiceId && !isCalendarDatesServiceId) {
+                    return Table.CALENDAR;
+                } else if (!isCalendarServiceId && isCalendarDatesServiceId) {
+                    return Table.CALENDAR_DATES;
+                } else {
+                    // For this case it doesn't seem to matter which is returned so going with calendar.
+                    defaultTable = Table.CALENDAR;
+                }
```
This could probably be simplified into
```java
if (!isCalendarServiceId && isCalendarDatesServiceId) {
    return Table.CALENDAR_DATES;
} else {
    return Table.CALENDAR;
}
```

br648 replied:
I've made changes to this in my latest update, needs further reviewing.

binh-dam-ibigroup commented on src/main/java/com/conveyal/datatools/manager/jobs/feedmerge/MergeLineContext.java:
```diff
@@ -237,42 +237,104 @@ public void startNewRow() throws IOException {
+            "%s:%s:%s",
+            table.name,
+            field.name,
+            field.referenceTables.stream().map(r -> r.name).collect(Collectors.joining(":"))
```
Hopefully we don't rely extract values from keys by relying on the ":" separator.

br648 replied:
I've made changes to this in my latest update to use a more unque key.

binh-dam-ibigroup reviewed (commented):
@br648 I think this is headed in the right direction! (Add the tests for checking that the logic works.)

Commit:
refactor(Multiple reference tables lookup): Feed merge updated to handle multi ref tables

binh-dam-ibigroup commented on src/main/java/com/conveyal/datatools/manager/jobs/feedmerge/ReferenceTableDiscovery.java:
```diff
@@ -0,0 +1,83 @@
+     * Define the reference table for a trip service id.
+     */
+    public static Table getTripServiceIdReferenceTable(
+        FieldContext fieldContext,
```
Make `fieldContext.getValueToWrite()` the first arg of the function, instead of `fieldContext`.

binh-dam-ibigroup reviewed (approved):
Works well when merging two normal feeds!

binh-dam-ibigroup commented:
@br648 Remember to address the merge conflict.

Commit:
refactor(fixed merge conflicts and addressed PR feedback):

Commit:
Merge branch 'dev' into feature/DT-183-Exception-based-scheduling

Commit:
refactor(Fixed merge conflict):

Commit:
refactor(pom.xml): Bumped gtfs lib to latest commit for exception based scheduling

philip-cline reviewed (approved):
This is looking good to me, although I think we should probably hold off on merging until we've got a front end fix for full testing.

philip-cline committed:
merge(dev): merge dev

philip-cline commented:
Awaiting https://github.com/ibi-group/datatools-server/pull/567 to be merged

br648 committed:
refactor(pom.xml): Bumped GTFS-Lib version

philip-cline committed:
Merge branch 'dev' into feature/DT-183-Exception-based-scheduling

philip-cline committed:
merge(dev): merge dev in again

philip-cline committed:
refactor(deps): update gtfs-lib commit

br648 committed:
refactor(ReferenceTableDiscovery.java): Updated trip service id key

key now includes reference to schedule exceptions matching trip service_id references
